File: rpc/RPCTransaction.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.rpc;

import com.google.protobuf.ByteString;
import grakn.client.Grakn.Transaction;
import grakn.client.GraknOptions;
import grakn.client.common.exception.GraknClientException;
import grakn.client.concept.ConceptManager;
import grakn.client.logic.LogicManager;
import grakn.client.query.QueryManager;
import grakn.protocol.GraknGrpc;
import grakn.protocol.TransactionProto;
import io.grpc.StatusRuntimeException;
import io.grpc.stub.StreamObserver;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.UUID;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Function;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static grakn.client.GraknProtoBuilder.options;
import static grakn.client.common.exception.ErrorMessage.Client.TRANSACTION_CLOSED;
import static grakn.client.common.exception.ErrorMessage.Client.UNKNOWN_REQUEST_ID;
import static grakn.client.common.tracing.TracingProtoBuilder.tracingData;
import static grakn.common.util.Objects.className;

public class RPCTransaction implements Transaction {

    private final Transaction.Type type;
    private final ConceptManager conceptManager;
    private final LogicManager logicManager;
    private final QueryManager queryManager;
    private final StreamObserver<TransactionProto.Transaction.Req> requestObserver;
    private final ResponseCollectors collectors;
    private final int networkLatencyMillis;
    private final AtomicBoolean isOpen;

    RPCTransaction(RPCSession.Core session, ByteString sessionId, Type type, GraknOptions options) {
        this.type = type;
        conceptManager = new ConceptManager(this);
        logicManager = new LogicManager(this);
        queryManager = new QueryManager(this);
        collectors = new ResponseCollectors();

        // Opening the StreamObserver exposes these atomics to another thread, so we must initialize them first.
        isOpen = new AtomicBoolean(true);
        requestObserver = GraknGrpc.newStub(session.channel()).transaction(responseObserver());
        final TransactionProto.Transaction.Req.Builder openRequest = TransactionProto.Transaction.Req.newBuilder()
                .putAllMetadata(tracingData())
                .setOpenReq(TransactionProto.Transaction.Open.Req.newBuilder()
                        .setSessionId(sessionId)
                        .setType(TransactionProto.Transaction.Type.forNumber(type.id()))
                        .setOptions(options(options)));
        final Instant startTime = Instant.now();
        final TransactionProto.Transaction.Open.Res res = execute(openRequest).getOpenRes();
        final Instant endTime = Instant.now();
        networkLatencyMillis = (int) ChronoUnit.MILLIS.between(startTime, endTime) - res.getProcessingTimeMillis();
    }

    @Override
    public Type type() {
        return type;
    }

    @Override
    public boolean isOpen() {
        return isOpen.get();
    }

    @Override
    public ConceptManager concepts() {
        return conceptManager;
    }

    @Override
    public LogicManager logic() {
        return logicManager;
    }

    @Override
    public QueryManager query() {
        return queryManager;
    }

    @Override
    public void commit() {
        final TransactionProto.Transaction.Req.Builder commitReq = TransactionProto.Transaction.Req.newBuilder()
                .putAllMetadata(tracingData())
                .setCommitReq(TransactionProto.Transaction.Commit.Req.getDefaultInstance());
        try {
            execute(commitReq);
        } finally {
            close();
        }
    }

    @Override
    public void rollback() {
        final TransactionProto.Transaction.Req.Builder rollbackReq = TransactionProto.Transaction.Req.newBuilder()
                .putAllMetadata(tracingData())
                .setRollbackReq(TransactionProto.Transaction.Rollback.Req.getDefaultInstance());
        execute(rollbackReq);
    }

    @Override
    public void close() {
        close(new Response.Done.Completed());
    }

    private void close(Response.Done doneResponse) {
        if (isOpen.compareAndSet(true, false)) {
            collectors.clear(doneResponse);
            requestObserver.onCompleted();
        }
    }

    public TransactionProto.Transaction.Res execute(TransactionProto.Transaction.Req.Builder request) {
        return executeAsync(request, res -> res).get();
    }

    public <T> QueryFuture<T> executeAsync(TransactionProto.Transaction.Req.Builder request, Function<TransactionProto.Transaction.Res, T> transformResponse) {
        if (!isOpen.get()) throw new GraknClientException(TRANSACTION_CLOSED);

        final ResponseCollector.Single responseCollector = new ResponseCollector.Single();
        final UUID requestId = UUID.randomUUID();
        request.setId(requestId.toString());
        collectors.put(requestId, responseCollector);
        requestObserver.onNext(request.build());
        return new QueryFuture<>(responseCollector, transformResponse);
    }

    public <T> Stream<T> stream(TransactionProto.Transaction.Req.Builder request, Function<TransactionProto.Transaction.Res, Stream<T>> transformResponse) {
        if (!isOpen.get()) throw new GraknClientException(TRANSACTION_CLOSED);

        final ResponseCollector.Multiple responseCollector = new ResponseCollector.Multiple();
        final UUID requestId = UUID.randomUUID();
        request.setId(requestId.toString());
        request.setLatencyMillis(networkLatencyMillis);
        collectors.put(requestId, responseCollector);
        requestObserver.onNext(request.build());
        final ResponseIterator<T> responseIterator = new ResponseIterator<>(requestId, requestObserver, responseCollector, transformResponse);
        return StreamSupport.stream(((Iterable<T>) () -> responseIterator).spliterator(), false);
    }

    private StreamObserver<TransactionProto.Transaction.Res> responseObserver() {
        return new StreamObserver<TransactionProto.Transaction.Res>() {
            @Override
            public void onNext(TransactionProto.Transaction.Res res) {
                if (!isOpen.get()) return;

                final UUID requestId = UUID.fromString(res.getId());
                final ResponseCollector collector = collectors.get(requestId);
                assert collector != null : UNKNOWN_REQUEST_ID.message(requestId);
                collector.add(new Response.Result(res));
            }

            @Override
            public void onError(Throwable ex) {
                assert ex instanceof StatusRuntimeException : "The server sent an exception of unexpected type " + ex;
                // TODO: this isn't nice - an error from one request isn't really appropriate for all of them (see #180)
                close(new Response.Done.Error((StatusRuntimeException) ex));
            }

            @Override
            public void onCompleted() {
                close(new Response.Done.Completed());
            }
        };
    }

    private static class ResponseCollectors {
        private final ConcurrentMap<UUID, ResponseCollector> collectors;

        private ResponseCollectors() {
            collectors = new ConcurrentHashMap<>();
        }

        private ResponseCollector get(UUID requestId) {
            return collectors.get(requestId);
        }

        private synchronized void put(UUID requestId, ResponseCollector collector) {
            collectors.put(requestId, collector);
        }

        private synchronized void clear(Response.Done doneResponse) {
            collectors.values().parallelStream().forEach(collector -> collector.add(doneResponse));
            collectors.clear();
        }
    }

    abstract static class ResponseCollector {
        private final AtomicBoolean isDone;
        /** gRPC response messages, errors, and notifications of abrupt termination are all sent to this blocking queue. */
        private final BlockingQueue<Response> responseBuffer;

        ResponseCollector() {
            isDone = new AtomicBoolean(false);
            responseBuffer = new LinkedBlockingQueue<>();
        }

        void add(Response response) {
            responseBuffer.add(response);
            if (!(response instanceof Response.Result) || isLastResponse(response.read())) isDone.set(true);
        }

        boolean isDone() {
            return isDone.get();
        }

        TransactionProto.Transaction.Res take() throws InterruptedException {
            return responseBuffer.take().read();
        }

        TransactionProto.Transaction.Res take(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
            final Response response = responseBuffer.poll(timeout, unit);
            if (response != null) return response.read();
            else throw new TimeoutException();
        }

        abstract boolean isLastResponse(TransactionProto.Transaction.Res response);

        static class Single extends ResponseCollector {

            @Override
            boolean isLastResponse(TransactionProto.Transaction.Res response) {
                return true;
            }
        }

        static class Multiple extends ResponseCollector {

            @Override
            boolean isLastResponse(TransactionProto.Transaction.Res response) {
                return response.getDone();
            }
        }
    }

    abstract static class Response {

        abstract TransactionProto.Transaction.Res read();

        static class Result extends Response {
            private final TransactionProto.Transaction.Res response;

            Result(TransactionProto.Transaction.Res response) {
                this.response = response;
            }

            @Override
            TransactionProto.Transaction.Res read() {
                return response;
            }

            @Override
            public String toString() {
                return className(getClass()) + "{" + response + "}";
            }
        }

        static abstract class Done extends Response {

            static class Completed extends Done {

                @Override
                TransactionProto.Transaction.Res read() {
                    throw new GraknClientException(TRANSACTION_CLOSED);
                }

                @Override
                public String toString() {
                    return className(getClass()) + "{}";
                }
            }

            static class Error extends Done {

                private final StatusRuntimeException  error;

                Error(StatusRuntimeException error) {
                    // TODO: parse different gRPC errors into specific GraknClientException
                    this.error = error;
                }

                @Override
                TransactionProto.Transaction.Res read() {
                    throw new GraknClientException(error);
                }

                @Override
                public String toString() {
                    return className(getClass()) + "{" + error + "}";
                }
            }
        }
    }
}

File: rpc/RPCTransaction.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
    Grakn,
    ConceptManager,
    ProtoBuilder,
    GraknOptions,
    QueryManager,
    uuidv4,
    BlockingQueue,
    Stream,
} from "../dependencies_internal";
import TransactionProto from "graknlabs-grpc-protocol/protobuf/transaction_pb";
import GraknProto from "graknlabs-grpc-protocol/protobuf/grakn_grpc_pb";
import GraknGrpc = GraknProto.GraknClient;
import { ClientDuplexStream } from "@grpc/grpc-js";

export class RPCTransaction implements Grakn.Transaction {
    private readonly _type: Grakn.TransactionType;
    private readonly _conceptManager: ConceptManager;
    private readonly _queryManager: QueryManager;
    private readonly _collectors: ResponseCollectors;
    private readonly _grpcClient: GraknGrpc;

    private _stream: ClientDuplexStream<TransactionProto.Transaction.Req, TransactionProto.Transaction.Res>;
    private _streamIsOpen: boolean;
    private _transactionWasOpened: boolean;
    private _transactionWasClosed: boolean;
    private _networkLatencyMillis: number;

    constructor(grpcClient: GraknGrpc, type: Grakn.TransactionType) {
        this._type = type;
        this._conceptManager = new ConceptManager(this);
        this._queryManager = new QueryManager(this);
        this._collectors = new ResponseCollectors(this);
        this._transactionWasClosed = false;
        this._transactionWasOpened = false;
        this._streamIsOpen = false;
        this._grpcClient = grpcClient;
    }

    async open(sessionId: string, options?: GraknOptions): Promise<RPCTransaction> {
        this.openTransactionStream();
        this._streamIsOpen = true;

        const openRequest = new TransactionProto.Transaction.Req()
            .setOpenReq(
                new TransactionProto.Transaction.Open.Req()
                    .setSessionId(sessionId)
                    .setType(this._type === Grakn.TransactionType.READ ? TransactionProto.Transaction.Type.READ : TransactionProto.Transaction.Type.WRITE)
                    .setOptions(ProtoBuilder.options(options))
            );
        const startTime = new Date().getTime();
        const res = await this.execute(openRequest, res => res.getOpenRes());
        const endTime = new Date().getTime();
        this._networkLatencyMillis = endTime - startTime - res.getProcessingTimeMillis();
        this._transactionWasOpened = true;
        return this;
    }

    public type(): Grakn.TransactionType {
        return this._type;
    }

    public isOpen(): boolean {
        return this._transactionWasOpened && !this._transactionWasClosed;
    }

    public concepts(): ConceptManager {
        return this._conceptManager;
    }

    public query(): QueryManager {
        return this._queryManager;
    }

    public async commit(): Promise<void> {
        const commitReq = new TransactionProto.Transaction.Req()
            .setCommitReq(new TransactionProto.Transaction.Commit.Req());
        await this.execute(commitReq);
    }

    public async rollback(): Promise<void> {
        const rollbackReq = new TransactionProto.Transaction.Req()
            .setRollbackReq(new TransactionProto.Transaction.Rollback.Req());
        await this.execute(rollbackReq);
    }

    async close(): Promise<void> {
        if (this._streamIsOpen) {
            this._streamIsOpen = false;
            // TODO: close stream, somehow?
        }
        if (!this._transactionWasClosed) {
            this._transactionWasClosed = true;
            this._collectors.clearWithError(new ErrorResponse("Transaction closed."))
        }
    }

    execute<T>(request: TransactionProto.Transaction.Req, transformResponse: (res: TransactionProto.Transaction.Res) => T = () => null): Promise<T> {
        const responseCollector = new ResponseCollector();
        const requestId = uuidv4();
        request.setId(requestId);
        this._collectors.put(requestId, responseCollector);
        // TODO: we can optionally inject the callback here - perhaps that would be cleaner than using ResponseCollectors?
        this._stream.write(request);
        return responseCollector.take().then(transformResponse);
    }

    stream<T>(request: TransactionProto.Transaction.Req, transformResponse: (res: TransactionProto.Transaction.Res) => T[]): Stream<T> {
        const responseCollector = new ResponseCollector();
        const requestId = uuidv4();
        request.setId(requestId);
        request.setLatencyMillis(this._networkLatencyMillis);
        this._collectors.put(requestId, responseCollector);
        this._stream.write(request);
        return new Stream<T>(requestId, this._stream, responseCollector, transformResponse);
    }

    private openTransactionStream() {
        this._stream = this._grpcClient.transaction();

        this._stream.on("data", (res) => {
            const requestId = res.getId();
            const collector = this._collectors.get(requestId);
            if (!collector) throw "Unknown request ID " + requestId;
            collector.add(new OkResponse(res));
        });

        this._stream.on("error", (err) => {
            console.error(err);
        });

        this._stream.on("end", () => {
            this._streamIsOpen = false;
            this.close();
        });
        // TODO: look into _stream.on(status) + any other events
    }
}

class ResponseCollectors {
    private readonly _map: { [requestId: string]: ResponseCollector };
    private readonly _transaction: RPCTransaction;
    constructor(transaction: RPCTransaction) {
        this._map = {};
        this._transaction = transaction;
    }

    get(uuid: string) {
        return this._map[uuid];
    }

    put(uuid: string, collector: ResponseCollector) {
        if (this._transaction["_transactionWasClosed"]) throw "The transaction has been closed and no further operation is allowed."
        this._map[uuid] = collector;
    }

    clearWithError(error: ErrorResponse) {
        Object.keys(this._map).forEach((requestId) => this._map[requestId].add(error));
        for (const requestId in this._map) delete this._map[requestId];
    }
}

export class ResponseCollector {
    private _responseBuffer: BlockingQueue<Response>;

    constructor() {
        this._responseBuffer = new BlockingQueue<Response>();
    }

    add(response: Response): void {
        this._responseBuffer.add(response);
    }

    async take(): Promise<TransactionProto.Transaction.Res> {
        const response = await this._responseBuffer.take();
        return response.read();
    }
}

abstract class Response {
    abstract read(): TransactionProto.Transaction.Res;
}

class OkResponse extends Response {
    private readonly _res: TransactionProto.Transaction.Res;

    constructor(res: TransactionProto.Transaction.Res) {
        super()
        this._res = res;
    }

    read(): TransactionProto.Transaction.Res {
        return this._res;
    }

    toString(): string {
        return "OkResponse {" + this._res.toString() + "}";
    }
}

class ErrorResponse extends Response {
    private readonly _error: Error | string;

    constructor(error: Error | string) {
        super();
        this._error = error;
    }

    read(): TransactionProto.Transaction.Res {
        throw this._error;
    }

    toString(): string {
        return "ErrorResponse {" + this._error.toString() + "}";
    }
}
